Spark: Add separate modules for Spark 3.1

wypoon opened the pull request:
Add a separate directory for v3.1 to spark and build and test against Spark 3.1 separately.
The reflection based code to enable cross-compilation against both Spark 3.0 and 3.1 are removed from the v3.1 code. They have not been removed from the v3.0 code; that will be done as a separate change.

rdblue commented:
@wypoon, can we split the PR into a copy and separate changes to remove reflection? I think that will make it much easier to track down what happened if we have to revisit the reflection changes.

rdblue commented on .github/workflows/spark-ci.yml:
```diff
@@ -99,6 +99,33 @@ jobs:
+        with:
+          name: test logs
+          path: |
+            **/build/testlogs
```
Instead of adding another job here, can we update the `matrix.spark` parameter in the `spark-32-tests` config? One of the reasons to add `3.2` to the module names instead of some derivation like `32` was so that we could build the module names from `matrix.spark`.

wypoon replied:
Ok, I'll try that. I'm not familiar with the CI. What is not clear to me is:
Suppose we have two values in
```
        spark: ['3.1', '3.2']
```
instead of the one value
```
        spark: ['3.1']
```
then in the
```
      - run: ./gradlew -DsparkVersions=${{ matrix.spark }} ...
```
does that get run *twice*, each time with one of the values? Or is that run once, with two values in `sparkValues` (because sparkVersions can take a comma-separated list of versions)?

rdblue replied:
Each run uses just one of the values in the build matrix. So for jvm: [8, 11] and spark: ['3.1', '3.2'] you'll get (8, 3.1), (11, 3.1), (8, 3.2), (11, 3.2).

wypoon commented:
> @wypoon, can we split the PR into a copy and separate changes to remove reflection? I think that will make it much easier to track down what happened if we have to revisit the reflection changes.

Sure, I can do that.

wypoon commented:
```
% diff -r spark/v3.0 spark/v3.1
diff -r spark/v3.0/build.gradle spark/v3.1/build.gradle
21,23c21,23
<     project(':iceberg-spark:iceberg-spark3'),
<     project(":iceberg-spark:iceberg-spark3-extensions"),
<     project(':iceberg-spark:iceberg-spark3-runtime')
---
>     project(':iceberg-spark:iceberg-spark-3.1'),
>     project(":iceberg-spark:iceberg-spark-3.1-extensions"),
>     project(':iceberg-spark:iceberg-spark-3.1-runtime')
28c28
<     sparkVersion = '3.0.3'
---
>     sparkVersion = '3.1.2'
41c41
< project(':iceberg-spark:iceberg-spark3') {
---
> project(':iceberg-spark:iceberg-spark-3.1') {
106c106
< project(":iceberg-spark:iceberg-spark3-extensions") {
---
> project(":iceberg-spark:iceberg-spark-3.1-extensions") {
132,133c132
<     compileOnly project(':iceberg-spark')
<     compileOnly project(':iceberg-spark:iceberg-spark3')
---
>     compileOnly project(':iceberg-spark:iceberg-spark-3.1')
146,147c145
<     testImplementation project(path: ':iceberg-spark', configuration: 'testArtifacts')
<     testImplementation project(path: ':iceberg-spark:iceberg-spark3', configuration: 'testArtifacts')
---
>     testImplementation project(path: ':iceberg-spark:iceberg-spark-3.1', configuration: 'testArtifacts')
163c161
< project(':iceberg-spark:iceberg-spark3-runtime') {
---
> project(':iceberg-spark:iceberg-spark-3.1-runtime') {
197,198c195,196
<     implementation project(':iceberg-spark:iceberg-spark3')
<     implementation project(':iceberg-spark:iceberg-spark3-extensions')
---
>     implementation project(':iceberg-spark:iceberg-spark-3.1')
>     implementation project(':iceberg-spark:iceberg-spark-3.1-extensions')
209,211c207,208
<     integrationImplementation project(path: ':iceberg-spark', configuration: 'testArtifacts')
<     integrationImplementation project(path: ':iceberg-spark:iceberg-spark3', configuration: 'testArtifacts')
<     integrationImplementation project(path: ':iceberg-spark:iceberg-spark3-extensions', configuration: 'testArtifacts')
---
>     integrationImplementation project(path: ':iceberg-spark:iceberg-spark-3.1', configuration: 'testArtifacts')
>     integrationImplementation project(path: ':iceberg-spark:iceberg-spark-3.1-extensions', configuration: 'testArtifacts')
213,214c210,211
<     integrationCompileOnly project(':iceberg-spark:iceberg-spark3-extensions')
<     integrationCompileOnly project(':iceberg-spark:iceberg-spark3')
---
>     integrationCompileOnly project(':iceberg-spark:iceberg-spark-3.1-extensions')
>     integrationCompileOnly project(':iceberg-spark:iceberg-spark-3.1')
253c250
<     description = "Test Spark3 Runtime Jar against Spark 3.0"
---
>     description = "Test Spark3 Runtime Jar against Spark 3.1"
diff -r spark/v3.0/spark/src/test/java/org/apache/iceberg/spark/data/TestSparkDateTimes.java spark/v3.1/spark/src/test/java/org/apache/iceberg/spark/data/TestSparkDateTimes.java
72c72
<     String sparkTimestamp = DateTimeUtils.timestampToString(formatter, ts.value());
---
>     String sparkTimestamp = formatter.format(ts.value());
iceberg % 
```

I rebased on master and updated v3.1. The only difference between v3.0 and v3.1 now is the build.gradle and TestSparkDateTimes.java. The change to the latter is needed because `DateTimeUtils.timestampToString` is no longer in Spark 3.1.
I also updated the CI.

rdblue reviewed (approved):
Looks great! I'll merge when tests are passing.

rdblue commented:
Thanks, @wypoon! Nice work.
